flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManagerImpl.java

```diff
@@ -767,8 +767,7 @@ private void internalRequestSlot(PendingSlotRequest pendingSlotRequest) throws R
 				// request can not be fulfilled by any free slot or pending slot that can be allocated,
 				// check whether it can be fulfilled by allocated slots
 				if (failUnfulfillableRequest && !isFulfillableByRegisteredSlots(pendingSlotRequest.getResourceProfile())) {
-					throw new ResourceManagerException("Could not fulfill slot request " + pendingSlotRequest.getAllocationId() + ". "
-						+ "Requested resource profile (" + pendingSlotRequest.getResourceProfile() + ") is unfulfillable.");
+					throw new UnfulfillableSlotRequestException(pendingSlotRequest);
```

Comment:
I think we also need to update the method `SlotManagerImpl#setFailUnfulfillableRequest` to fail the slots with the proper exception. Please also add a test because this is apparently not guarded.